server/remote_cache/byte_stream_server/byte_stream_server.go

```diff
@@ -402,6 +394,59 @@ func (s *ByteStreamServer) QueryWriteStatus(ctx context.Context, req *bspb.Query
 	}, nil
 }
 
+func handleAlreadyExists(stream bspb.ByteStream_WriteServer, firstRequest *bspb.WriteRequest) error {
+	r, err := digest.ParseUploadResourceName(firstRequest.ResourceName)
+	if err != nil {
+		return err
+	}
+	// The following code implements a workaround for the way Bazel 5.0.0 handles
+	// writing compressed blobs.
+	//
+	// When the server closes the stream and sends the WriteResponse, Bazel will
+	// check whether the committed_size matches the number of bytes it has
+	// uploaded, but only if it has uploaded all bytes. But because Bazel doesn't
+	// wait for an ack before it uploads each chunk, we (generally) have no way of
+	// knowing whether Bazel has uploaded all bytes at this point. So we are
+	// forced to wait until bazel uploads all chunks, so that we know for sure
+	// that the committed_size we send back will match Bazel's expectation.
+	//
+	// There are two cases where we can short-circuit, though:
+	//
+	// - If this is an uncompressed stream, we assume that the committed_size will
```

Comment:
probably worth comment what's happening in the biggest part of the function below: if this is a compressed stream?

Reply:
Reworded and shortened -- hopefully it's clearer now.